Add History to Memory Viewer

Vancleeff opened the pull request:
Enhanced user experience by providing an history on the memory viewer with the last 50 explored addresses (can move backward or forward)

Vancleeff committed:
Add History to Memory Viewer
Enhanced user experience by providing an history on the memory viewer with the last 50 explored addresses (can move backward or forward)

Jamiras commented on src/ui/viewmodels/MemoryViewerViewModel.hh:
```diff
@@ -235,6 +239,8 @@ private:
     static std::unique_ptr<ra::ui::drawing::ISurface> s_pFontASCIISurface;
     static int s_nFont;
+    static std::list<ra::ByteAddress> s_listMemViewHistory;
+    static std::list<ra::ByteAddress>::iterator s_iterMemViewHistoryIndex;
```
`std::vector` is generally preferred over `std::list` if you're not inserting stuff in the middle. `std::list` uses node pointers to make updates faster at the cost of memory. `ra::ByteAddress` is a 32-bit value, so the overhead of the node is quite expensive.

Also, there's no reason for these to be static. You wouldn't want your history shared across multiple instances - And I'm not really sure how this will interact with the pointer finder. You might be okay since you're managing the addition of items to the history from the inspector.

Jamiras commented on src/ui/win32/bindings/MemoryViewerControlBinding.cpp:
```diff
@@ -57,6 +57,13 @@ INT_PTR CALLBACK MemoryViewerControlBinding::WndProc(HWND hControl, UINT uMsg, W
+        case WM_XBUTTONUP:
+            if (GET_XBUTTON_WPARAM(wParam) == XBUTTON1)
+                m_pViewModel.MoveMemViewHistoryBackward();
+            else
```
Recommend explicitly checking for XBUTTON2 in case other XBUTTONs ever get defined.

Jamiras commented on src/ui/viewmodels/MemoryViewerViewModel.hh:
```diff
@@ -175,6 +175,10 @@ public:
 
+    void SaveToMemViewHistory();
+    void MoveMemViewHistoryBackward();
+    void MoveMemViewHistoryForward();
```
`MemView` is redundant in these function names as they're part of the `MemoryViewerViewModel`.

Jamiras commented on src/ui/viewmodels/MemoryViewerViewModel.hh:
```diff
@@ -175,6 +175,10 @@ public:
     void AdvanceCursorPage();
     void RetreatCursorPage();
 
+    void SaveToMemViewHistory();
```
Suggest `AddToHistory`, saving implies something that gets remembered between sessions.

Might also want a `ClearHistory` method so the addresses can be reset when changing games.

Jamiras commented on src/ui/viewmodels/MemoryViewerViewModel.cpp:
```diff
@@ -1314,6 +1317,33 @@ void MemoryViewerViewModel::RenderHeader()
+        s_listMemViewHistory.erase(s_listMemViewHistory.begin(), s_iterMemViewHistoryIndex);
+        if (s_listMemViewHistory.size() > MEMVIEW_HISTORY_MAX_SIZE)
+            s_listMemViewHistory.pop_back();
+        s_listMemViewHistory.push_front(GetAddress());
```
Recommend pushing new items on the back to avoid reallocating data until the history is full.

Jamiras commented on src/ui/viewmodels/MemoryViewerViewModel.cpp:
```diff
@@ -1314,6 +1317,33 @@ void MemoryViewerViewModel::RenderHeader()
+    {
+        s_listMemViewHistory.erase(s_listMemViewHistory.begin(), s_iterMemViewHistoryIndex);
+        if (s_listMemViewHistory.size() > MEMVIEW_HISTORY_MAX_SIZE)
+            s_listMemViewHistory.pop_back();
```
I think you might have an off-by-one error here. If `size() > MEMVIEW_HISTORY_MAX_SIZE`, that means there's 51 items in the list. Even if you pop the 51st item, you're about to put another 51st back on.

Jamiras commented on src/ui/viewmodels/MemoryViewerViewModel.hh:
```diff
@@ -235,6 +239,8 @@ private:
     static std::unique_ptr<ra::ui::drawing::ISurface> s_pFontASCIISurface;
     static int s_nFont;
+    static std::list<ra::ByteAddress> s_listMemViewHistory;
+    static std::list<ra::ByteAddress>::iterator s_iterMemViewHistoryIndex;
```
Suggest `s_iterMemViewHistoryIndex` just be an integer. Maintaining a iterator over a dynamically sized array can be difficult. Also, I think it would improve readability.

vancleeff replied:
I switch to vector and refactor the logic on the last commit. I also tried to switch to integer index instead of iterator but I found it way less readable than the iterator version.

There's probably a better way to do it so I post it that version here and let you judge which version should be implemented and/or what to improve : 

```
void MemoryViewerViewModel::AddToHistory(ra::ByteAddress nAddress)
{
    if (vHistory.at(nHistoryIndex) != nAddress)
    {
        vHistory.erase(std::next(vHistory.begin(), nHistoryIndex + 1), vHistory.end());
        if (vHistory.size() == MEMVIEW_HISTORY_MAX_SIZE)
            vHistory.erase(vHistory.begin());
        vHistory.push_back(nAddress);
        nHistoryIndex = vHistory.size() - 1;
    }
}

void MemoryViewerViewModel::MoveHistoryForward()
{
    int nIndex = std::distance(vHistory.begin(), std::next(vHistory.begin(), nHistoryIndex));
    if (vHistory.size() > (nIndex + 1))
        ++nHistoryIndex;
    SetAddress(vHistory.at(nHistoryIndex));
}

void MemoryViewerViewModel::MoveHistoryBackward()
{
    if (vHistory.size() >= 1 and std::next(vHistory.begin(), nHistoryIndex) != vHistory.begin())
        --nHistoryIndex;
    SetAddress(vHistory.at(nHistoryIndex));
}
```

VS
```
void MemoryViewerViewModel::AddToHistory(ra::ByteAddress nAddress)
{
    if (*iterHistoryIndex != nAddress)
    {
        vHistory.erase(iterHistoryIndex + 1, vHistory.end());
        if (vHistory.size() == MEMVIEW_HISTORY_MAX_SIZE)
            vHistory.erase(vHistory.begin());
        vHistory.push_back(nAddress);
        iterHistoryIndex = vHistory.end() - 1;
    }
}

void MemoryViewerViewModel::MoveHistoryForward()
{
    int nIndex = std::distance(vHistory.begin(), iterHistoryIndex);
    if (vHistory.size() > (nIndex + 1))
        ++iterHistoryIndex;
    SetAddress(*iterHistoryIndex);
}

void MemoryViewerViewModel::MoveHistoryBackward()
{
    if (vHistory.size() >= 1 and iterHistoryIndex != vHistory.begin())
        --iterHistoryIndex;
    SetAddress(*iterHistoryIndex);
}
```

Jamiras commented on src/ui/viewmodels/MemoryInspectorViewModel.cpp:
```diff
@@ -287,6 +287,7 @@ void MemoryInspectorViewModel::OnCurrentAddressChanged(ra::ByteAddress nNewAddre
     SetValue(CurrentAddressValueProperty, nValue);
 
     m_pViewer.SetAddress(nNewAddress);
+    m_pViewer.SaveToMemViewHistory();
```
I was originally going to suggest you just capture the new address in `SetAddress`, but putting it here addresses another concern I had about the address being captured when navigating via the keyboard. Recommend having unit tests that ensure the history is not modified by keyboard navigation.

vancleeff replied:
I moved the AddToHistory directly inside SetAddress() as suggested. 
Concerning capturing the keyboard navigation, I was thinking about that at some point and I found it way more convenient to capture it, it's too confusing if we don't track these movements : 

Let's say we are on 0x0 then we move to 0x123 after which we start to use the keyboard to find something interresting somewhere near 0x123, after some digging we want to come back to 0x123 so we MoveBackward() but we reach 0x0. 
If we want to get back to 0X123 we need to MoveForward() now, which is confusing.

Jamiras replied:
You don't want to capture keyboard movements. If you jump to address $1234, then press right and down to get to $1245, then type "ABCDEF", you'll be at $1248 and your history will be: [1234->1235->1245->1246->1247->1248]. If I went back in this scenario, I'd expect to jump back to $1234 (the last thing I jumped _to_), not to the last character I typed.

Testing in Visual Studio, it's a little more complicated. It doesn't remember where you jump to, it remembers where you jump from. If you click on row 6, character 8, then row 12, character 10, and press down and type "banana", when you hit back it goes to row 6, character 8. Then forward takes you to the position right after "banana".

Vancleeff committed:
Refactor history feature following review

 - Update naming
 - Move the history list to vector
 - Add ClearHistory function
 - Refactor of the AddToHistory logic
 - Add unit tests

vancleeff commented:
Committed ! Thanks @Jamiras !

Vancleeff commented on src/ui/viewmodels/MemoryViewerViewModel.cpp:
```diff
@@ -94,6 +95,8 @@ class MemoryViewerViewModel::MemoryBookmarkMonitor : protected ViewModelCollecti
     ViewModelCollection<ra::ui::viewmodels::MemoryBookmarksViewModel::MemoryBookmarkViewModel>& m_vBookmarks;
     MemoryViewerViewModel& m_pOwner;
+    std::vector<ra::ByteAddress> vHistory;
+    std::vector<ra::ByteAddress>::iterator iterHistoryIndex;
```
My bad, not needed indeed, thanks !

Jamiras commented:
This has conflicts from the +/- PR.
